explainer/effect-generic-bounds-and-functions.md

```diff
@@ -0,0 +1,544 @@
+`postgres` crate [[sync][pg], [async][pga]], as well as the `reqwest` crate
+[[sync][rqw], [async][rqwa]], and both the [`tokio`] and [`async-std`] crates
+duplicate large swaths of the stdlib's functionality. Other effects are also
+covered, such as [`fallible-iterator`] for a version of the stdlib's `Iterator`
+trait which short-circuits on error.  And [`fallible_vec`] for a `Vec` type with
+methods which may returns errors rather than panics if an allocation fails.
+
+[`rspotify`]: https://docs.rs/rspotify/latest/rspotify/index.html
+[mdb]: https://docs.rs/mongodb/latest/mongodb/sync/index.html
+[mdba]: https://docs.rs/mongodb/latest/mongodb/index.html
+[pg]: https://docs.rs/postgres/latest/postgres/
+[pga]: https://docs.rs/tokio-postgres/latest/tokio_postgres/index.html
+[rqw]: https://docs.rs/reqwest/latest/reqwest/blocking/index.html
+[rqwa]: https://docs.rs/reqwest/latest/reqwest/
+[`tokio`]: https://docs.rs/tokio
+[`async-std`]: https://docs.rs/async-std
+[`fallible-iterator`]: https://docs.rs/fallible-iterator/latest/fallible_iterator/index.html
+[`fallible_vec`]: https://crates.io/crates/fallible_vec
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Effect-generic functions
+
+Effects such as `async`, `try`, or `gen` define a superset of the language. With
+some minor exceptions, they provide access to more features than
+functions which don't have the effect. However, to ensure the effect forwards
+correctly through function bodies, we require some degree of annotations. In the
+case of the `async` effect, we require function calls to be suffixed with
+`.await`. In the case of `try`, we require function calls to be suffixed with
+`?`. This is called "effect forwarding".
+
+Say we wanted to write a function `sum` which takes an `impl Iterator<Item =
+u32>` and sums all numbers together. If we included the trait definition, we could write it like so:
+
+```rust
+/// The `Iterator` trait
+pub trait Iterator {
+    type Item;
+    fn next(&mut self) -> Option<Self::Item>;
+}
+
+/// Iterate over all numbers in the
+/// iterator and sum them together
+pub fn sum<I: Iterator<Item = u32>>(iter: I) -> u32 {
+    let mut total = 0;
+    while let Some(n) = iter.next() {
+        total += n;
+    }
+    total
+}
+```
+
+This works fine for non-async code. In fact: this is almost exactly how the
+stdlib versions of
+[`Iterator::sum`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.sum)
+is defined. But this code has a limitation: the underlying iterator will always
+block between calls to `next`. To resolve that, `next` should add support for
+`async/.await`, which we can do by adding the `#[maybe(async)]` notation.
+
+```rust
+/// The `Iterator` trait with optional
+/// support for the `async` effect
+#[maybe(async)]
+pub trait Iterator {
+    type Item;
+    #[maybe(async)]
+    fn next(&mut self) -> Option<Self::Item>;
```

Comment:
what happens when you want a pinned `async Iterator`:

```rust
trait AsyncIterator {
    type Item;
    fn poll_next(self: Pin<&mut Self>) -> Poll<Option<Self::Item>>;
}
```

afaict using `async` in this style makes that difficult.

Reply:
The signature for pinned iterators under this model would be:

```rust
#[maybe(async)]
pub trait PinnedIterator {
    type Item;
    #[maybe(async)]
    fn next(self: Pin<&mut Self>) -> Option<Self::Item>;
```

This not only covers the async case, but also the non-async case. Pinning however has its own set of problems. Pin projections are an essential part of the pinning system, but exist outside of the type system. So ideally we'd be able to enable functions to be polymorphic over the pinning guarantees too - enabling us to write "maybe pin" functions or similar. But that's its own topic which we shouldn't get into on this PR.